test/yjsTestEditor.ts

```diff
@@ -0,0 +1,152 @@
+    const remoteEvents = events.filter((event) => !event.transaction.local);
+    if (remoteEvents.length == 0) {
+      return;
+    }
+
+    e.isRemote = true;
+
+    Editor.withoutNormalizing(e, () => {
+      toSlateOps(remoteEvents).forEach((op) => {
+        e.apply(op);
+      });
+    });
+
+    Promise.resolve().then(() => (e.isRemote = false));
+  },
+
+  /**
+   * Capture Yjs updates generated by this editor.
+   */
+  captureYjsUpdate: (e: YjsTestEditor, update: Uint8Array, _origin: any) => {
+    if (!e.shouldCaptureYjsUpdates) return;
+    e.capturedYjsUpdates.push(update);
+  },
+
+  /**
+   * Return captured Yjs updates.
+   */
+  getCapturedYjsUpdates: (e: YjsTestEditor): Uint8Array[] => {
+    const result = e.capturedYjsUpdates;
+    e.capturedYjsUpdates = [];
+    return result;
+  },
+
+  /**
+   * Apply one Yjs update to Yjs.
+   */
+  applyYjsUpdateToYjs: (
+    e: YjsTestEditor,
+    update: Uint8Array
+  ): Promise<void> => {
+    return new Promise((resolve) => {
+      e.shouldCaptureYjsUpdates = false;
+      e.onChangeComplete = () => {
+        e.onChangeComplete = () => void {};
+        resolve();
+      };
+      Y.applyUpdate(e.doc, update);
+      e.shouldCaptureYjsUpdates = true;
+    });
+  },
+
+  /**
+   * Apply multiple Yjs updates to Yjs.
+   */
+  applyYjsUpdatesToYjs: async (e: YjsTestEditor, updates: Uint8Array[]) => {
+    await Promise.all(
+      updates.map((update) => {
+        YjsTestEditor.applyYjsUpdateToYjs(e, update);
+      })
+    );
+  },
+
+  /**
+   * Apply one slate operation to slate.
+   */
+  applySlateOpToSlate: (e: YjsTestEditor, op: Operation): Promise<void> => {
+    return new Promise((resolve) => {
+      e.onChangeComplete = () => {
+        e.onChangeComplete = () => void {};
+        resolve();
+      };
+      e.apply(op);
+    });
+  },
+
+  /**
+   * Apply multiple slate operations to slate.
+   */
+  applySlateOpsToSlate: async (e: YjsTestEditor, operations: Operation[]) => {
+    await Promise.all(
+      operations.map((op) => {
+        YjsTestEditor.applySlateOpToSlate(e, op);
+      })
+    );
+  },
+};
+
+export const withYjsTest = <T extends Editor>(editor: T): T & YjsTestEditor => {
+  const e = editor as T & YjsTestEditor;
+
+  const doc = new Y.Doc();
+  const syncDoc = doc.getArray<SyncElement>('content');
+
+  doc.on('update', (updateMessage: Uint8Array, origin: any) => {
+    YjsTestEditor.captureYjsUpdate(e, updateMessage, origin);
+  });
+  syncDoc.observeDeep((events) => {
+    YjsTestEditor.applyYjsEventsToSlate(e, events);
+  });
+
+  e.doc = doc;
+  e.syncDoc = syncDoc;
+  e.isRemote = false;
+  e.shouldCaptureYjsUpdates = true;
+  e.capturedYjsUpdates = [];
+
+  const { onChange } = editor;
+  e.onChange = () => {
+    if (!e.isRemote) {
+      YjsTestEditor.applySlateOpsToYjs(e, e.operations);
+    }
+
+    if (onChange) {
+      onChange();
+    }
+
+    if (e.onChangeComplete) {
+      e.onChangeComplete();
+    }
```

Comment:
Same here, you can just wrap the yjs editor onChange to call onChangeComplete and remove all of this code duplication.

Reply:
Oh yeah, I totally overlooked that. Makes sense.